### PR TITLE
Fixed crash during log loading

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
This PR, proposes to use the AsyncTask to load the log, this help us to see the log problem during the debugging also if the logging is too much.

Also, this use the line reader, this can help us in the future to load a partial log and not all.

I hope this can help.

### CODE REVIEW

File: app/src/main/java/com/lvaccaro/lamp/LogActivity.kt [hunk 1]
```diff
@@ -54,37 +59,57 @@ class LogActivity : AppCompatActivity() {
 
     fun readLog() {
         title = "Log $daemon"
-        val logFile = File(rootDir(),"$daemon.log")
+
+        val logFile = File(rootDir(), "$daemon.log")
+        Log.d(TAG, "File log: ${logFile.absolutePath}")
+        Log.d(TAG, "File dim: ${logFile.length()} Mb")
+        Log.d(TAG, "------------------------------------------")
+        Log.d(TAG, logFile.readText())
```
Comment (lvaccaro) [inline app/src/main/java/com/lvaccaro/lamp/LogActivity.kt:67]: this read the full file and print in the logger, I think it is not good for big file.

File: app/src/main/java/com/lvaccaro/lamp/LogActivity.kt [hunk 2]
```diff
@@ -54,37 +59,57 @@ class LogActivity : AppCompatActivity() {
...
-                val line = logReader.readLine() ?: break
-                sb.append(line)
-            } catch (e: Exception) {
-                break
+    private class LoadLogTask(val activity: LogActivity, val editText: EditText): AsyncTask<File, String, String>() {
+
+        override fun doInBackground(vararg params: File?): String? {
+            var text: String
+            var logReader = LineNumberReader(params[0]?.reader())
+            var lines: List<String> = logReader.lines().collect(Collectors.toList())
```
Comment (lvaccaro) [inline app/src/main/java/com/lvaccaro/lamp/LogActivity.kt:90]: the file is full read in memory at 1 time: it is really slow (it should be better an incremental read) and memory consuming, moreover it is copied in text variable, so consumed the memory twice the size of the file.

File: app/src/main/java/com/lvaccaro/lamp/LogActivity.kt [hunk 3]
```diff
@@ -54,37 +59,57 @@ class LogActivity : AppCompatActivity() {
...
 
-    fun read100(logReader: BufferedReader): String {
-        val sb = StringBuilder()
-        for (i in 0..100) {
-            try {
-                val line = logReader.readLine() ?: break
-                sb.append(line)
-            } catch (e: Exception) {
-                break
+    private class LoadLogTask(val activity: LogActivity, val editText: EditText): AsyncTask<File, String, String>() {
```
Comment (lvaccaro) [inline app/src/main/java/com/lvaccaro/lamp/LogActivity.kt:85]: nit: you could also use `doAsync { }` & `runOnMainThread {}` instead `AsyncTask`

Comment (vincenzopalazzo) [reply app/src/main/java/com/lvaccaro/lamp/LogActivity.kt:85]: I will do tomorrow, thanks to improve my code

File: app/src/main/java/com/lvaccaro/lamp/LogActivity.kt (hunk 1 above)
Comment (vincenzopalazzo) [reply app/src/main/java/com/lvaccaro/lamp/LogActivity.kt:67]: mh! yep.

I use this to debug and I didn't think about this thing.

I will remove it tomorrow!

File: app/src/main/java/com/lvaccaro/lamp/LogActivity.kt (hunk 2 above)
Comment (vincenzopalazzo) [reply app/src/main/java/com/lvaccaro/lamp/LogActivity.kt:90]: You have right, I think this method needs a little bit more work on the design.

**Comment by vincenzopalazzo** [comment]
Hi @lvaccaro,

After several hours of debugging of this situation because for some reason I see this crash also with the async task, I stopped to work and I start to write some code with the last version of coroutine of kotlin, to see if I can reproduce the bug also with this thing, and I can.

I discovered that the bug was how much text we were added inside the Edit text. For some reason, the text sometimes was too much and the UI Thread don't have a lot of time to update the UI in time.

To prove this I implement the log activity with the coroutines and ViewModel architectural pattern only to test my idea and the result on [this branch](https://github.com/vincenzopalazzo/lamp/tree/refactoring_log_view)

This is the result



The conclusion are the following:

- The bug is not inside the previous implementation, it is only the method that we add the text inside the EditText. I found a solution like this

https://github.com/vincenzopalazzo/lamp/blob/630f4429614086a93515842926cc8152a590de19/app/src/main/java/com/lvaccaro/lamp/logview/LogViewModel.kt#L67-L85

- The ViewModel and Couritines add a little more complexity (or better, they add rules) inside the code but the code is a little cleaner (I found somethings where kotlin is more cool than java :) ).
  - I don't know that you want to add these things inside Lamp at the moment, In fact, is possible to add my idea to the previous implementation, maybe we lose the real loading, but not is enough important see the log that will update to runtime in the view.


The things that I want to add inside this PR is

- [ ] Fixed crash of app is the text is too much (Fixed, we should be decide only the implementation)
- [ ] Understand because sometimes the log is in hexadecimal
- [ ] Add the button tu share the log with email or telegram

**Comment by lvaccaro** [comment]
Thank for all the testing. 

An alternative solution could be to write in the EditText only a small part of the log (the last ~200 lines): I usually read the log to detect crash and node failures. As you suggested, the full file will be downloaded by sharing. 
Maybe the hexadecimal log part is related to the esplora plugin, it shows blockhash and block when it fetches. I agree it is too much overhead.

**Comment by vincenzopalazzo** [comment]
Hi @lvaccaro,

> An alternative solution could be to write in the EditText only a small part of the log (the last ~200 lines)

This is my goal! but I'm thinking about how to store how many the line contains the file because the reading operation is useful to do with only one time in the life of the app on phone. If we can store the dimension of file in someplace we can try to view a sample code to load only the 200 lines inside the log file. and with this information, we can maintain the last lines of the log update inside the text field easily. At the moment my solution is to use the SharePropieties to store the position of file from LogObserver and use this info inside the LogView.

My main problem with this solution is, how I can remove the last line inside the "edit text" and update only the last line in the "edit text"? is possible to implement the solution with a FIFO structure but the problem how to update the EditTex on in partial and not in toto, any suggestion about this?

>Maybe the hexadecimal log part is related to the esplora plugin, it shows blockhash and block when it fetches. I agree it is too much overhead.

I can try to make a test in the next day, about esplora log

**Comment by lvaccaro** [comment]
Thanks, esplora plugin is too much verbose, it should be update to be more quite. If you want, you could open an issue on esplora plugin repo.

I don't know if some approach here could be useful:
https://stackoverflow.com/questions/4121678/java-read-last-n-lines-of-a-huge-file

**Comment by vincenzopalazzo** [comment]
>Thanks, esplora plugin is too much verbose, it should be update to be more quite. If you want, you could open an issue on esplora plugin repo.

Yesterday, I start to work with esplora plugin and lightningd, I try it and maybe when I have a complete test I can make some issue about it.

>I don't know if some approach here could be useful:
https://stackoverflow.com/questions/4121678/java-read-last-n-lines-of-a-huge-file

In my opinion, this is a cool idea, use something to reverse the file or to random access, I'm thinking if exist some fault in this solution. For instance, if I will open the log during the lightningd are writing on it how can use resolve this problem?

**Comment by vincenzopalazzo** [comment]
Closed by https://github.com/lvaccaro/lamp/pull/47

**Pull request closed without merging**